image/cas/layout/tar.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"archive/tar"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/opencontainers/image-spec/image/cas"
+	"github.com/opencontainers/image-spec/image/layout"
+)
+
+// TarEngine is a cas.Engine backed by a tar file.
+type TarEngine struct {
+	reader ReadSeekCloser
+}
+
+// NewTarEngine returns a new TarEngine.
+func NewTarEngine(file ReadSeekCloser) (eng cas.Engine, err error) {
+	engine := &TarEngine{
+		reader: file,
+	}
+
+	err = layout.CheckTarVersion(engine.reader)
+	if err != nil {
+		return nil, err
+	}
+
+	return engine, nil
+}
+
+// Put adds a new blob to the store.
+func (engine *TarEngine) Put(reader io.Reader) (digest string, err error) {
+	// FIXME
+	return "", errors.New("TarEngine.Put is not supported yet")
+}
+
+// Get returns a reader for retrieving a blob from the store.
+func (engine *TarEngine) Get(digest string) (reader io.ReadCloser, err error) {
+	fields := strings.SplitN(digest, ":", 2)
+	if len(fields) != 2 {
+		return nil, fmt.Errorf("invalid digest: %q, %v", digest, fields)
+	}
+	algorithm := fields[0]
+	hash := fields[1]
+
+	targetName := fmt.Sprintf("./blobs/%s-%s", algorithm, hash)
+
+	_, err = engine.reader.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return nil, err
+	}
+
+	tarReader := tar.NewReader(engine.reader)
+	for {
+		header, err := tarReader.Next()
```

Comment:
On Thu, Jul 07, 2016 at 02:57:35PM -0700, Stephen Day wrote:
> > +	tarReader := tar.NewReader(engine.reader)
> > +	for {
> > +		header, err := tarReader.Next()
>
> Rather than doing an `O(N)` search through the tar every time, it is
> sufficient to make a single pass over the tar, noting the offset and
> length of the target file. When searching for the file in Get, this
> becomes and index lookup and seek, rather than having to scan the
> whole file.

I'm not really worried about performance here [1].  Iterating each
time is the simplest approach that works, so I think it's a good match
for the spec repo's reference implementation.  Folks who intend to use
the Engine interfaces in production will probably want to pick up an
external implementation aimed at performance.  If I was using this in
production I would certainly not be using an implementation based on
tar ;).

That said, I'm happy to put in straightforward optimizations if they
wouldn't distract from the reference-implementation goal.

[1]: https://github.com/opencontainers/image-spec/pull/159/files#r70000054